File: test-vulnerabilities/typescript/sql-injection-test.ts
/**
 * SQL Injection 脆弱性テスト用ファイル
 * 
 * 検出対象: Semgrep (p/typescript, カスタムルール)
 * 脆弱性: CWE-89 SQL Injection
 */

interface Database {
    query(sql: string): Promise<any[]>;
    execute(sql: string, params?: any[]): Promise<any>;
}

// 脆弱性1: 文字列連結によるSQL Injection
async function getUserByName(db: Database, name: string): Promise<any> {
    // NG: 文字列連結でSQLを構築
    const sql = "SELECT * FROM users WHERE name = '" + name + "'";
    return await db.query(sql);
}

// 脆弱性2: テンプレートリテラルによるSQL Injection  
async function searchProducts(db: Database, keyword: string, category: string): Promise<any[]> {
    // NG: テンプレートリテラルで直接埋め込み
    const sql = `SELECT * FROM products 
                 WHERE name LIKE '%${keyword}%' 
                 AND category = '${category}'`;
    return await db.query(sql);
}

// 脆弱性3: 動的テーブル名（SQL Injection）
async function getTableData(db: Database, tableName: string): Promise<any[]> {
    // NG: テーブル名を直接埋め込み
    const sql = `SELECT * FROM ${tableName}`;
    return await db.query(sql);
}

// 安全な実装例（参考）
async function getUserByNameSafe(db: Database, name: string): Promise<any> {
    // OK: パラメータ化クエリを使用
    const sql = "SELECT * FROM users WHERE name = ?";
    return await db.execute(sql, [name]);
}

export { getUserByName, searchProducts, getTableData, getUserByNameSafe };

File: test-vulnerabilities/java/SqlInjectionTest.java
package com.yourco.web.vulnsamples;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;

import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

/**
 * SQL Injection 脆弱性テスト用クラス
 *
 * 検出対象: Semgrep (p/java, カスタムルール)
 * 脆弱性: CWE-89 SQL Injection
 */
@RestController
@RequestMapping("/test/sqli")
public class SqlInjectionTest {

    private Connection connection;

    /**
     * 脆弱性1: 文字列連結によるSQL Injection
     * 修正方法: PreparedStatement を使用
     */
    @GetMapping("/users")
    public String getUsers(@RequestParam String name) throws SQLException {
        // NG: 文字列連結 - SQL Injection 脆弱性
        String sql = "SELECT * FROM users WHERE name = '" + name + "'";
        Statement stmt = connection.createStatement();
        ResultSet rs = stmt.executeQuery(sql);
        return rs.toString();
    }

    /**
     * 脆弱性2: String.format によるSQL Injection
     * 修正方法: PreparedStatement を使用
     */
    @GetMapping("/orders")
    public String getOrders(@RequestParam String userId) throws SQLException {
        // NG: String.format - SQL Injection 脆弱性
        String sql = String.format("SELECT * FROM orders WHERE user_id = '%s'", userId);
        Statement stmt = connection.createStatement();
        ResultSet rs = stmt.executeQuery(sql);
        return rs.toString();
    }

    /**
     * 安全な実装例（参考）
     */
    @GetMapping("/safe/users")
    public String getUsersSafe(@RequestParam String name) throws SQLException {
        // OK: パラメータバインディング (PreparedStatement)
        String sql = "SELECT * FROM users WHERE name = ?";
        PreparedStatement pstmt = connection.prepareStatement(sql);
        pstmt.setString(1, name);
        ResultSet rs = pstmt.executeQuery();
        return rs.toString();
    }
}

